### PR TITLE
✨ Autoscaling: buffer machines Episode II (⚠️ devops)

### PR DESCRIPTION
**Pull request by sanderegg**
## What do these changes do?
continues from #3790 

This PR reworked the machine buffering enhancement. After this PR the autoscaling service can keep a constant number of so-called "reserve machines" that are in docker swarm *drain* mode. Therefore these machines are ready for running services in a few seconds (max 20 seconds which is the time the autoscaling background task interval + possibly starting other instances).

#### Use-cases:
1. no load: the reserved drain machines are kept forever
2. under load: when machines in the reserve are used for services, then new machines are pre-created and added to the reserve again

#### Next steps:
- implement some heuristic to have *smart reserve*. Why keep unneeded resources forever if there is a long time of innactivity?
-  (⚠️ devops) ```EC2_INSTANCES_MACHINES_BUFFER``` is now usable and shall be defined as ```0``` to disable machine buffering, or any number to enable it. 
- NOTE: The type of machine used for buffering is the first one defined in ```EC2_INSTANCES_ALLOWED_TYPES```

More details:
- a new model that objectify the concept of a cluster enables much easier handling

### CODE REVIEW

File: services/autoscaling/src/simcore_service_autoscaling/models.py [hunk 1]
```diff
@@ -9,6 +9,31 @@
 from types_aiobotocore_ec2.literals import InstanceStateNameType, InstanceTypeType
 
 
+class SimcoreServiceDockerLabelKeys(BaseModel):
+    # NOTE: in a next PR, this should be moved to packages models-library and used
+    # all over, and aliases should use io.simcore.service.*
+    # https://github.com/ITISFoundation/osparc-simcore/issues/3638
```
Comment (pcrespov) [inline services/autoscaling/src/simcore_service_autoscaling/models.py:15]: NOTE: we have this functionality as well in the service integration library . There I think is pretty solid. I agree that we should somehow put all this code in a single place. I will add some hints in the issue above

File: services/autoscaling/src/simcore_service_autoscaling/models.py [hunk 2]
```diff
@@ -84,3 +85,33 @@ class EC2InstanceData:
...
+        metadata={
+            "description": "This is an EC2 instance that is not yet associated to a docker node"
+        }
+    )
+    disconnected_nodes: list[Node] = field(
+        metadata={
+            "description": "This is a docker node which is not backed by a running EC2 instance"
+        }
+    )
+    terminated_instances: list[EC2InstanceData]
```
Comment (pcrespov) [inline services/autoscaling/src/simcore_service_autoscaling/models.py:117]: MINOR metadata.description here?
Q: Is this dataclass field metadata used somewhere in the code (e.g. to generate specs/reports... or just for doc ?

File: services/autoscaling/src/simcore_service_autoscaling/models.py [hunk 3]
```diff
@@ -37,37 +62,13 @@ def __sub__(self, other: "Resources") -> "Resources":
         )
 
 
-class EC2Instance(BaseModel, frozen=True):
+@dataclass(frozen=True)
+class EC2InstanceType:
```
Comment (pcrespov) [inline services/autoscaling/src/simcore_service_autoscaling/models.py:66]: So this is information of the ec2 instance (even before it is started)  and EC2InstaceData is extra/other information of the instance once is started, right?

**Review by pcrespov — APPROVED** [review]
Great job! spotless!

File: services/autoscaling/src/simcore_service_autoscaling/models.py (hunk 3 above)
Comment (sanderegg) [reply services/autoscaling/src/simcore_service_autoscaling/models.py:66]: yes

File: services/autoscaling/src/simcore_service_autoscaling/models.py (hunk 2 above)
Comment (sanderegg) [reply services/autoscaling/src/simcore_service_autoscaling/models.py:117]: no that is currently documentation only. I like having this in the pydantic models, now I can do the same in the dataclasses when these are complex objects.

Comment (pcrespov) [reply services/autoscaling/src/simcore_service_autoscaling/models.py:117]: THOUGHT: i always wondered what is the impact of these in memory vs a simple comment provided that they are used for doc.

File: services/autoscaling/src/simcore_service_autoscaling/utils/dynamic_scaling.py [hunk 4]
```diff
@@ -105,18 +107,21 @@ async def try_assigning_task_to_pending_instances(
...
+            now = datetime.datetime.now(datetime.timezone.utc)
+            time_since_launch = now - instance.launch_time
+            estimated_time_to_completion = (
+                instance.launch_time + _MAX_TIME_TO_START_EC2_INSTANCE - now
+            )
             await log_tasks_message(
                 app,
                 [pending_task],
-                "scaling up of cluster in progress...awaiting new machines...please wait...",
+                f"adding machines to the cluster (time waiting: {time_since_launch}, est. remaining time: {estimated_time_to_completion})...please wait...",
```
Comment (GitHK) [inline services/autoscaling/src/simcore_service_autoscaling/utils/dynamic_scaling.py:118]: nice!

File: services/autoscaling/tests/unit/test_dynamic_scaling_core.py [hunk 5]
```diff
@@ -276,35 +444,20 @@ async def test_cluster_scaling_up(
...
+        instance_state="running",
+    )
 
     # as the new node is already running, but is not yet connected, hence not tagged and drained
-    mock_try_get_node_with_name.assert_not_called()
+    mock_find_node_with_name.assert_not_called()
     mock_tag_node.assert_not_called()
     mock_set_node_availability.assert_not_called()
+    mock_compute_node_used_resources.assert_not_called()
+    # mock_cluster_used_resources.assert_called_once()
```
Comment (GitHK) [inline services/autoscaling/tests/unit/test_dynamic_scaling_core.py:460]: is this correct here? it is commented out

**Review by GitHK — APPROVED** [review]
👍

**Pull request merged**